include/srf/experimental/modules/segment_modules.hpp

```diff
@@ -0,0 +1,171 @@
+ */
+
+#pragma once
+
+#include "nlohmann/json.hpp"
+
+#include "srf/segment/object.hpp"
+
+#include <map>
+#include <string>
+#include <typeinfo>
+#include <utility>
+#include <vector>
+
+namespace srf::segment {
+class Builder;
+}
+
+namespace srf::modules {
+
+class SegmentModule
+{
+  public:
+    using segment_module_port_map_t     = std::map<std::string, std::shared_ptr<segment::ObjectProperties>>;
+    using segment_module_port_t         = std::shared_ptr<segment::ObjectProperties>;
+    using segment_module_typeinfo_map_t = std::map<std::string, const std::type_info*>;
+
+    SegmentModule() = delete;
+    SegmentModule(std::string module_name);
+    SegmentModule(std::string module_name, nlohmann::json config);
+
+    const std::string& component_prefix() const;
+    const nlohmann::json& config() const;
+    const std::string& name() const;
+
+    std::string get_module_component_name(const std::string& component_name) const;
+
+    /**
+     * Return vector of input names -- these are only understood by the SegmentModule
+     * @return std::vector
+     */
+    const std::vector<std::string>& input_ids() const;
+
+    /**
+     * Return a vector of output names -- these are only understood by the SegmentModule class
+     * @return std::vector
+     */
+    const std::vector<std::string>& output_ids() const;
+
+    /**
+     * Return a set of ObjectProperties for module input_ids
+     * @return ObjectProperties
+     */
+    const segment_module_port_map_t& input_ports() const;
+
+    /**
+     * Return the ObjectProperties object corresponding to input_name
+     * @param input_name Name of the module port
+     * @return ObjectProperties
+     */
+    segment_module_port_t input_port(const std::string& input_name);
+
+    /**
+     * Return a map of module port id : type indices
+     * @return std::map
+     */
+    const std::map<std::string, const std::type_info*>& input_port_type_ids() const;
+
+    /**
+     * Return the type index of a given input name
+     * @return Type index
+     */
+    const std::type_info* input_port_type_id(const std::string& input_name);
+
+    /**
+     * Return a set of ObjectProperties for module input_ids
+     * @return ObjectProperties
+     */
+    const segment_module_port_map_t output_ports() const;
+
+    /**
+     * Return an ObjectProperties for the module port corresponding to output name
+     * @param output_name Name of the module port to return
+     * @return ObjectProperties
+     */
+    segment_module_port_t output_port(const std::string& output_name);
+
+    /**
+     * Return a map of module port id : type indices
+     * @return std::map
+     */
+    const segment_module_typeinfo_map_t& output_port_type_ids() const;
+
+    /**
+     * Return the type index of a given input name
+     * @param input_name Name of the module port to return a type index for
+     * @return Type index
+     */
+    const std::type_info* output_port_type_id(const std::string& output_name);
```

Comment:
could this just return a `const std::type_info&` instead of a pointers?  
should this method be `const`?

same holds for the `input` variant

Reply:
No, the issue is essentially the same explanation as to why I didn't use a type_index, its an issue with map's `[]` operator behavior wrt to values that can't be default constructed. It looks like there are some work arounds to get the desired access behavior along with utilizing either a type_info& or a type_index, but it seems like a bit overkill.

Another option could be an indirect lookup, where the map just keeps an offset into a vector that holds a type_index.

For example:
```c++
std::map<std::string, const std::type_info&> m_map;

const std::type_info& get_thing(std::string key) {
    return m_map[key]; // error: In template: reference to type 'const std::type_info' requires an initializer
}
```

Reply:
Since we store type_index everywhere else I'll move to using that for consistency.